### PR TITLE
[FLINK-24851][Connectors / Kafka] KafkaSourceBuilder: auto.offset.reset is ignored

### PR DESCRIPTION
**Pull request by hililiwei**
## What is the purpose of the change

Creating KafkaSource like this:
```

val props = new Properties()
props.put("bootstrap.servers", "localhost:9092")
props.put("group.id", "group1")
props.put("auto.offset.reset", "latest")
val kafkaSource = KafkaSource.builder[String]()
      .setProperties(props)
      .build()
```
The actually used value for "auto.offset.reset" is "earliest" instead of configured "latest".
This occurs because "auto.offset.reset" gets overridden by startingOffsetsInitializer.getAutoOffsetResetStrategy().name().toLowerCase(). The default value for startingOffsetsInitializer is "earliest".

This behavior is misleading.

This behavior imposes an inconvenience on configuring the Kafka connector. We cannot use the Kafka setting "auto.offset.reset" as-is. Instead we must extract this particular setting from other settings and propagate to KafkaSourceBuilder.setStartingOffsets():

```
val kafkaSource = KafkaSource.builder[String]()
      .setProperties(props)
      .setStartingOffsets(
        OffsetsInitializer.committedOffsets(
          OffsetResetStrategy.valueOf(
            props.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG)
              .asInstanceOf[String]
              .toUpperCase(Locale.ROOT)
          )
        )
      )
      .build()
```
The expected behavior is to use the value of "auto.offset.reset" provided by KafkaSourceBuilder.setProperties() - unless overridden via KafkaSourceBuilder. setStartingOffsets().

https://issues.apache.org/jira/browse/FLINK-24851

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by hililiwei** [comment]
cc @AHeise could you please help to review the code? thank you.

**Comment by AHeise** [comment]
@PatrickRen can you PTAL?

**Comment by PatrickRen** [comment]
Thanks for the patch and sorry for my late response!

Basically I think this is not a bug and I left some comments in [FLINK-24851](https://issues.apache.org/jira/browse/FLINK-24851). Feel free to continue the discussion in the ticket~

**Comment by hililiwei** [comment]
@PatrickRen can you PTAL?

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java [hunk 1]
```diff
@@ -438,9 +457,13 @@ private void parseAndSetRequiredProperties() {
             maybeOverride(KafkaSourceOptions.COMMIT_OFFSETS_ON_CHECKPOINT.key(), "false", false);
         }
         maybeOverride(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false", false);
+
         maybeOverride(
                 ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,
-                startingOffsetsInitializer.getAutoOffsetResetStrategy().name().toLowerCase(),
+                startingOffsetsInitializer
```
Comment (tashoyan) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java:463]: Could be NPE? You removed the initialization for `startingOffsetsInitializer` from the constructor.

UPDATE: I see, that `sanityCheck()` is called before `parseAndSetRequiredProperties()` - and this allows us to escape from NullPointerException. However, this approach seems very fragile to me. I suggest to set `startingOffsetsInitializer` in `parseAndSetRequiredProperties()`.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java [hunk 2]
```diff
@@ -497,6 +520,29 @@ private void sanityCheck() {
                         "Property %s is required when offset commit is enabled",
                         ConsumerConfig.GROUP_ID_CONFIG));
         // Check offsets initializers
+        if (startingOffsetsInitializer == null) {
+            if (props.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG) != null) {
+                startingOffsetsInitializer =
+                        OffsetsInitializer.committedOffsets(
```
Comment (tashoyan) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java:526]: I believe, the function `sanityCheck()` is for checking the validity of the settings, but not to change them. Shouldn't we instead update the settings in `parseAndSetRequiredProperties()`?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 3]
```diff
@@ -355,6 +352,24 @@ public void run() {
...
+        producerProperties.putAll(secureProps);
+        KafkaProducer<String, String> kafkaProducer = new KafkaProducer<>(producerProperties);
+
+        for (int i = 0; i < parallelism; i++) {
+            for (int l = 0; l < recordsInEachPartition; l++) {
+                kafkaProducer.send(new ProducerRecord<>(topicName, i, "", String.valueOf(l)));
+            }
+        }
+        kafkaProducer.close();
+
```
Comment (tashoyan) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:372]: What is the purpose of this change? How is it related to the changes in KafkaSourceBuilder and KafkaSinkBuilder?

Comment (hililiwei) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:372]: https://github.com/apache/flink/blob/ca72f9e9bd95ad8a5368089d7bc0367256339b8b/flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java#L310-L312

After modifying the code, the test case failed to run. I checked its code and found that the test case didn't throw any exceptions before modifying the code, but it didn't seem to work as expected. It uses properties to set the value of auto.offset.reset to latest, and this value will be overwritten as early, which leads to the fact that the data obtained by the consumer has actually arrived in advance. Therefore, it is inconsistent with the above description. 

In order to make this test case achieve the expected effect, I start sending data to topic after the consumer starts, and test whether the consumer can obtain the data in the expected way.

If I am wrong, please correct. Thank you.

https://github.com/apache/flink/blob/ca72f9e9bd95ad8a5368089d7bc0367256339b8b/flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java#L324-L337

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java (hunk 2 above)
Comment (hililiwei) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java]: Modified, thank you.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 4]
```diff
@@ -355,6 +352,24 @@ public void run() {
                 };
         runner.start();
 
+        Thread.sleep(2 * 1000L);
```
Comment (tashoyan) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:355]: Is this an accepted way of thread synchronization in the Kafka community?
`Thread.sleep()`
Generally, it is a poor practice:
1. It is impossible to find a good delay. The real delay depends on many factors: hardware, OS, running processes, GC pause, etc.
2. The test is flaky - sometimes it passes, sometimes it fails
3. There is a temptation to increase the delay - this makes the test slow. If many tests use this practice of Thread.sleep(), then all the tests take considerably longer time. Nevertheless, sometimes this or that test fails.

Instead we can just check the properties of a `KafkaSink` after it has been built with the `KafkaSinkBuilder`. This should be enough to test whether `KafkaSinkBuilder` is implemented correctly.

Comment (hililiwei) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:355]: Agree with you, here, as elsewhere, is poorly written. It is to wait flink task to start. I'll modify it and use the return of `getRunningJobs(client)`.

**Comment by hililiwei** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
@hililiwei The Flink Kafka connector resides in it's own repository nowadays, if this code change is still relevant, please open the PR in https://github.com/apache/flink-connector-kafka
